### PR TITLE
A basic implementation of a3c

### LINKED ISSUE #21: Implement Actor-Critic
@article{konda1999actor,
  title={Actor-critic algorithms},
  author={Konda, Vijay and Tsitsiklis, John},
  journal={Advances in neural information processing systems},
  volume={12},
  year={1999}
}

https://proceedings.neurips.cc/paper_files/paper/1999/file/6449f44a102fde848669bdd9eb6b76fa-Paper.pdf

### PR DESCRIPTION
**Pull request by Tom-Pecher**
Closes #21 

Please look over the modifications to actor.py and critic.py before merging.

[a2c-episode-999.meta.json](https://github.com/user-attachments/files/18199318/a2c-episode-999.meta.json)

### CODE REVIEW

File: src/agents/a2c/agent.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        self.training = False
+        self.total_steps = 0
+
+    def select_action(self, state):
+        state = torch.FloatTensor(state).unsqueeze(0).to(self.device)
+        with torch.no_grad():
+            action = self.actor(state)
+            # Add exploration noise during training
+            if self.training:
+                noise = torch.randn_like(action) * 0.1
```
Comment (TheGogy) [inline src/agents/a2c/agent.py]:
```suggestion
                noise = torch.randn_like(action) * self.noise_level
```

File: src/agents/a2c/agent.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+class A2CAgent:
+    def __init__(self, state_dim, action_dim, max_action, device):
+        self.device = device
+        self.max_action = max_action
+        self.action_dim = action_dim
+
+        # Training hyperparameters
+        self.gamma = 0.99
+        self.lr = 3e-4
+        self.entropy_coef = 0.01
```
Comment (TheGogy) [inline src/agents/a2c/agent.py:22]:
```suggestion
        self.entropy_coef = 0.01
        self.noise_level = 0.1
```

File: src/agents/a2c/agent.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        returns = []
+        R = 0
+        for reward, done in zip(reversed(rewards), reversed(dones)):
+            if done:
+                R = 0
+            R = reward + self.gamma * R
+            returns.insert(0, R)
+        return torch.tensor(returns, dtype=torch.float32).to(self.device)
+
+    def train(self, trajectories):
```
Comment (TheGogy) [inline src/agents/a2c/agent.py:57]: Might be an idea to use the replay_buffer class?
Possibly using something similar to the other actor-critic agents (ddpg, etc)

File: src/agents/a2c/agent.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        states, actions, rewards, dones, _ = trajectories
+        self.training = True
+
+        # Convert to tensors
+        states = torch.FloatTensor(np.array(states)).to(self.device)
+        actions = torch.FloatTensor(np.array(actions)).to(self.device)
+        returns = self.compute_returns(rewards, dones)
```
Comment (TheGogy) [inline src/agents/a2c/agent.py:64]: This updates from the entire list every time the loss updates. Might be an idea to update this incrementally?

File: src/agents/a2c/train.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    max_action = float(env_info['action_high'][0])  # Get max_action from environment
+
+    agent = A2CAgent(state_dim, action_dim, max_action, device)  # Pass max_action to agent
+
+    best_reward = float('-inf')
+
+    for episode in range(num_episodes):
+        state, _ = env.reset()
+        episode_reward = 0
+        states, actions, rewards, dones = [], [], [], []
```
Comment (TheGogy) [inline src/agents/a2c/train.py:64]: Could definitely use the memory buffer here

File: src/agents/common/actor.py [hunk 6]
```diff
@@ -18,5 +18,11 @@ def __init__(self, state_dim, action_dim, max_action):
             nn.Tanh()
         )
 
+        # Initialize weights
```
Comment (TheGogy) [inline src/agents/common/actor.py:21]: Agents might want to initialize the weights differently - might be an idea to move this out of the `common/actor.py` file and put it into the agent initialization directly?

This would allow us to be more flexible in future when we want to copy agents from previous runs.

File: src/agents/common/critic.py [hunk 7]
```diff
@@ -17,5 +17,12 @@ def __init__(self, state_dim, action_dim):
...
+        # Initialize weights
+        for m in self.modules():
+            if isinstance(m, nn.Linear):
+                nn.init.orthogonal_(m.weight)
+                nn.init.constant_(m.bias, 0)
+
     def forward(self, state, action):
-        return self.network(torch.cat([state, action], 1))
+        x = torch.cat([state, action], dim=-1)
+        return self.network(x)
```
Comment (TheGogy) [inline src/agents/common/critic.py]:
```suggestion
```

File: src/agents/common/critic.py [hunk 8]
```diff
@@ -17,5 +17,12 @@ def __init__(self, state_dim, action_dim):
...
 
+        # Initialize weights
+        for m in self.modules():
+            if isinstance(m, nn.Linear):
+                nn.init.orthogonal_(m.weight)
+                nn.init.constant_(m.bias, 0)
+
     def forward(self, state, action):
-        return self.network(torch.cat([state, action], 1))
+        x = torch.cat([state, action], dim=-1)
```
Comment (TheGogy) [inline src/agents/common/critic.py]:
```suggestion
        return self.network(torch.cat([state, action], 1))
```

**Review by TheGogy — CHANGES_REQUESTED** [review]
LGTM! A few modifications here and there but overall good.
